service/src/main/java/bio/terra/axonserver/service/cromwellworkflow/CromwellWorkflowService.java

```diff
@@ -0,0 +1,186 @@
+            end,
+            status,
+            name,
+            id,
+            label,
+            labelor,
+            excludeLabelAnd,
+            excludeLabelOr,
+            additionalQueryResultFields,
+            includeSubworkflows);
+  }
+
+  /** Retrieve the labels of a workflow. */
+  public CromwellApiLabelsResponse getLabels(UUID workflowId)
+      throws bio.terra.cromwell.client.ApiException {
+    return new WorkflowsApi(getApiClient())
+        .labels(CROMWELL_CLIENT_API_VERSION, workflowId.toString());
+  }
+
+  /**
+   * Checks if the workflow has the required workspace id label (e.g.,
+   * "terra-workspace-id:workspaceId").
+   */
+  private void validateWorkflowLabelMatchesWorkspaceId(UUID workflowId, UUID workspaceId)
+      throws bio.terra.cromwell.client.ApiException {
+    Map<String, String> labels = getLabels(workflowId).getLabels();
+    if (labels.get(WORKSPACE_ID_LABEL_KEY) == null
+        || !labels.get(WORKSPACE_ID_LABEL_KEY).equals(workspaceId.toString())) {
+      throw new ApiException(
+          "Workflow %s is not a member of workspace %s".formatted(workflowId, workspaceId));
+    }
+  }
+
+  /**
+   * Check if the user has workspace access and if the workflow has the required workspace id label
+   * (e.g., "terra-workspace-id:workspaceId").
+   *
+   * @param workflowId identifier of the workflow
+   * @param workspaceId workspace where the workflow located
+   * @param accessToken access token
+   * @throws bio.terra.cromwell.client.ApiException Exception thrown by Cromwell client.
+   */
+  public void validateWorkspaceAccessAndWorkflowLabelMatches(
+      UUID workflowId, UUID workspaceId, String accessToken)
+      throws bio.terra.cromwell.client.ApiException {
+    // Check workspace access.
+    wsmService.checkWorkspaceReadAccess(workspaceId, accessToken);
+    // Then check if the workflow has the corresponding workspace id.
+    validateWorkflowLabelMatchesWorkspaceId(workflowId, workspaceId);
+  }
+
+  public ApiWorkflowQueryResponse toApiQueryResponse(
+      CromwellApiWorkflowQueryResponse workflowQuery) {
+    List<ApiWorkflowQueryResult> results =
+        workflowQuery.getResults().stream()
+            .map(
+                r ->
+                    new ApiWorkflowQueryResult()
+                        .id(r.getId())
+                        .name(r.getName())
+                        .status(r.getStatus())
+                        .submission(r.getSubmission())
+                        .start(r.getStart())
+                        .end(r.getEnd()))
+            .toList();
+
+    return new ApiWorkflowQueryResponse()
+        .results(results)
+        .totalResultsCount(workflowQuery.getTotalResultsCount());
+  }
+
+  public ApiWorkflowMetadataResponse toApiMetadataResponse(
+      CromwellApiWorkflowMetadataResponse metadataResponse) {
+    ObjectMapper mapper = new ObjectMapper();
+    return mapper.convertValue(metadataResponse, ApiWorkflowMetadataResponse.class);
+  }
```

Comment:
How should we be converting between the API models? 

Do we allow Jackson to deserialize and parse (since the models are nearly identical), or individually set the fields (as done above for query response)?

Reply:
I think individually makes sense to me.